src/chain.js

```diff
@@ -11,19 +11,25 @@ var map = require('./map');
  *
  * Dispatches to the `chain` method of the second argument, if present,
  * according to the [FantasyLand Chain spec](https://github.com/fantasyland/fantasy-land#chain).
+ * Treats the second argument as a FantasyLand Chain, when the the second argument is a function
+ * of the form `a -> x -> b`.
```

Comment:
This program does not compile:

```haskell
module Main where

import Control.Monad ((>>=))

append :: a -> [a] -> [a]
append x xs = xs ++ [x]

chain :: Monad m => (a -> m b) -> m a -> m b
chain f m = m >>= f

f = chain head append

main = return ()
```

```console
$ ghc Chain.hs
[1 of 1] Compiling Main             ( Chain.hs, Chain.o )

Chain.hs:11:16:
    Couldn't match type ‘[a -> b]’ with ‘[a] -> [a]’
    Expected type: a -> [a -> b]
      Actual type: a -> [a] -> [a]
    Relevant bindings include f :: a -> b (bound at Chain.hs:11:1)
    Probable cause: ‘append’ is applied to too few arguments
    In the second argument of ‘chain’, namely ‘append’
    In the expression: chain head append
```

I believe there's a type error in `R.chain(R.head, R.append)` but it somehow "works" anyway. I don't see how to get from `Chain m => (a -> m b) -> m a -> m b` to the signature you're proposing. Do you?

This program compiles:

```haskell
module Main where

import Control.Monad ((>>=))

append :: a -> [a] -> [a]
append x xs = xs ++ [x]

chain :: Monad m => (a -> m b) -> m a -> m b
chain f m = m >>= f

f = chain append head

main = putStrLn $ show $ f [1,2,3]
```

It prints `[1,2,3,1]` when run. :)

It looks as though there's a bug in this code:

> ```javascript
> return function() {
>   return monad.call(this, fn.apply(this, arguments)).apply(this, arguments);
> };
> ```